chain/network/src/routing/graph_v2/mod.rs

```diff
@@ -0,0 +1,668 @@
+use crate::network_protocol::{AdvertisedPeerDistance, Edge, EdgeState};
+use crate::routing::edge_cache::EdgeCache;
+use crate::routing::routing_table_view::RoutingTableView;
+use crate::stats::metrics;
+use arc_swap::ArcSwap;
+use near_async::time;
+use near_primitives::network::PeerId;
+use parking_lot::Mutex;
+use std::collections::VecDeque;
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+#[cfg(not(test))]
+use crate::concurrency;
+#[cfg(not(test))]
+use rayon::iter::ParallelBridge;
+
+#[cfg(test)]
+mod testonly;
+#[cfg(test)]
+mod tests;
+
+pub type NextHopTable = HashMap<PeerId, Vec<PeerId>>;
+
+#[derive(Clone)]
+pub struct GraphConfigV2 {
+    pub node_id: PeerId,
+    pub prune_edges_after: Option<time::Duration>,
+}
+
+pub enum NetworkTopologyChange {
+    PeerConnected(PeerId, Edge),
+    PeerDisconnected(PeerId),
+    PeerAdvertisedDistances(network_protocol::DistanceVector),
+}
+
+/// Locally stored properties of a received network_protocol::DistanceVector message
+struct PeerDistances {
+    /// Advertised distances indexed by the local EdgeCache's peer to id mapping.
+    pub distance: Vec<i32>,
+    /// The lowest nonce among all edges used to validate the distances.
+    /// For simplicity, used to expire the entire dstance vector at once.
+    pub min_nonce: u64,
+}
+
+struct Inner {
+    config: GraphConfigV2,
+
+    /// Data structure maintaing information about the entire known network
+    edge_cache: EdgeCache,
+
+    /// Edges of the local node's direct connections
+    local_edges: HashMap<PeerId, Edge>,
+    /// Distances advertised by the local node's direct peers
+    peer_distances: HashMap<PeerId, PeerDistances>,
+
+    /// Distances from the local node to other nodes
+    my_distances: HashMap<PeerId, u32>,
+    /// The latest DistanceVector advertised by the local node
+    my_distance_vector: network_protocol::DistanceVector,
+}
+
+impl Inner {
+    /// Function which verifies signed edges.
+    /// Returns true iff all the edges provided were valid.
+    ///
+    /// This method implements a security measure against an adversary sending invalid edges.
+    /// It verifies edges in parallel until the first invalid edge is found. It adds nonces
+    /// for all the edges verified so far to the cache, but drops all the remaining ones. This way
+    /// the wasted work (verification of invalid edges) is constant, no matter how large the input
+    /// size is.
+    ///
+    /// Edge verification is expensive, and it would be an attack vector if we dropped on the
+    /// floor valid edges verified so far: an attacker could prepare a message containing
+    /// a lot of valid edges, except for the last one, and send it repeatedly to a node.
+    /// The node would then validate all the edges every time, then reject the whole set
+    /// because just the last edge was invalid. Instead, we cache all the edges verified so
+    /// far and return an error only afterwards.
+    #[cfg(not(test))]
```

Comment:
why tho

Reply:
To avoid conflicting with the version of this function in `graph_v2/testonly.rs` [(link)](https://github.com/near/nearcore/pull/9187/files/f50201d62848ee770dde5bab6a084ef35ad54029#diff-35a31b9a89c89d33af89b903af70ea6946203ef57ed7f203a1ae9d8ce4bd576eR12-R18)